tests/unit_tests/lcls_tools/common/devices/test_device.py

```diff
@@ -0,0 +1,266 @@
+    ) -> None:
+        good_pv_name = "SOLN:GUNB:212:BCTRL"
+        good_pv_obj = PV(good_pv_name)
+        mock_get_attribute.return_value = good_pv_obj
+        result = self.device._get_pv_object_from_str("bact")
+        self.assertIsNotNone(result)
+        self.assertEqual(result, good_pv_obj)
+
+    def test_get_pv_object_from_str_with_bad_pv_string(self):
+        result = self.device._get_pv_object_from_str("bad-name")
+        self.assertIsNone(result)
+
+    @patch(
+        "lcls_tools.common.devices.device.Device._get_attribute",
+        new_callable=MagicMock(),
+    )
+    def test_get_callbacks_with_good_pv_string(
+        self,
+        mock_get_attribute: MagicMock,
+    ) -> None:
+        good_pv_name = "SOLN:GUNB:212:BCTRL"
+        good_pv_obj = PV(good_pv_name)
+        handle = "bact"
+        mock_get_attribute.return_value = good_pv_obj
+        result = self.device.get_callbacks(handle)
+        self.assertIsNotNone(result)
+        # start with empty dictionary check
+        self.assertEqual(result, dict())
+
+        def dummy_callback():
+            print("test callback")
+
+        # add some callback function to the pv
+        self.device.add_callback_to_pv(pv=handle, function=dummy_callback)
+        result = self.device.get_callbacks(handle)
+        # check we have an indexed-dictionary of callback functions
+        expected_callback_structure = {1: (dummy_callback, {})}
+        self.assertEqual(result, expected_callback_structure)
+
+    def test_get_callbacks_with_bad_pv_string(self):
+        result = self.device.get_callbacks("bad-pv-handle")
+        self.assertIsNone(result)
+
+    def test_get_callback_index_with_no_callbacks_for_pv(self):
+        mock_pv = self.pv_obj
+
+        def mock_callback():
+            print("mock callback")
+
+        index = self.device._get_callback_index(mock_pv, mock_callback)
+        self.assertIsNone(index)
+
+    @patch(
+        "lcls_tools.common.devices.device.Device._get_attribute",
+        new_callable=MagicMock(),
+    )
+    def test_get_callback_index_with_multiple_callbacks_for_pv(
+        self,
+        mock_get_attribute: MagicMock,
+    ) -> None:
+        mock_pv = self.pv_obj
+        mock_get_attribute.return_value = mock_pv
+
+        def mock_callback(message: str) -> None:
+            print(f"callback: {message}")
+
+        first_callback = lambda: mock_callback("first")
+        second_callback = lambda: mock_callback("second")
+        third_callback = lambda: mock_callback("third")
+        self.device.add_callback_to_pv(
+            "bact",
+            first_callback,
+        )
+        self.device.add_callback_to_pv(
+            "bact",
+            second_callback,
+        )
+        self.device.add_callback_to_pv(
+            "bact",
+            third_callback,
+        )
+        index = self.device._get_callback_index(
+            mock_pv,
+            second_callback,
+        )
+        self.assertEqual(index, 2)
```

Comment:
comments

Reply:
Added some comments for the more opaque tests in test_device.py

Reply:
Thanks!